File: php/src/Speakap/SDK/SignedRequest.php
<?php

namespace Speakap\SDK;

use \Speakap\Date\ExtendedDateTime;
use Speakap\SDK\Exception\ExpiredSignatureException;
use Speakap\SDK\Exception\InvalidSignatureException;

class SignedRequest
{
    /**
     * The default window a request is valid, in seconds
     */
    const DEFAULT_WINDOW = 60;

    /**
     * A RFC3986 encoded string
     * @var string
     */
    private $payload;

    /**
     * The query string as elements
     * @var array
     */
    private $decodedPayload;

    /**
     * @var string
     */
    private $appId;

    /**
     * The shared secret
     * @var string
     */
    private $appSecret;

    /**
     * The time window in seconds
     * @var int
     */
    private $signatureWindowSize;

    /**
     * @param string $appId
     * @param string $appSecret
     * @param int    $signatureWindowSize Time in seconds that the window is considered valid
     */
    public function __construct($appId, $appSecret, $signatureWindowSize = null)
    {
        $this->appId = $appId;
        $this->appSecret = $appSecret;
        $this->signatureWindowSize = $signatureWindowSize === null ? static::DEFAULT_WINDOW : $signatureWindowSize;
    }

    /**
     * Set the raw payload. Typically used in conjunction with isValid()
     *
     * @param string $payload typically set via: file_get_contents('php://input');
     *
     * @throws \InvalidArgumentException
     * @return $this
     */
    public function setPayload($payload)
    {
        $payloadProperties = $this->decodePayloadToArray($payload);
        if ( ! $this->isValidPayload($payloadProperties)) {
            throw new \InvalidArgumentException('Missing payload properties, got: '. print_r($payloadProperties, true));
        }

        $this->payload = $payload;
        $this->decodedPayload = $payloadProperties;

        return $this;
    }

    /**
     * @param array $params
     *
     * @throws \InvalidArgumentException
     *
     * @return bool
     */
    public function validateSignature(array $params)
    {
        if ( ! $this->isValidPayload($params)) {
            throw new \InvalidArgumentException('Missing payload properties, got: '. print_r($params, true));
        }

        if ($params['signature'] !== $this->getSignatureFromParameters($this->appSecret, $params)) {
            throw new InvalidSignatureException('Invalid signature, got: '. print_r($params, true));
        }

        $issuedAt = ExtendedDateTime::createFromFormat(\DateTime::ISO8601, $params['issuedAt']);
        if ( ! $this->isWithinWindow($this->signatureWindowSize, $issuedAt)) {
            throw new ExpiredSignatureException('Expired signature, got: '. print_r($params, true));
        }

        return true;
    }

    /**
     * Whether or not the payload is valid
     *
     * @return boolean
     */
    public function isValid()
    {
        if ($this->payload !== $this->getSelfSignedRequest($this->appSecret, $this->decodedPayload)) {
            // The payload doesn't match
            return false;
        }

        $issuedAt = ExtendedDateTime::createFromFormat(\DateTime::ISO8601, $this->decodedPayload['issuedAt']);
        if ( ! $this->isWithinWindow($this->signatureWindowSize, $issuedAt)) {
            // The date of the request does not fall in the allowed window size.
            return false;
        }

        return true;
    }

    /**
     * This method returns an encoded signed request. It does not re-use the original input,
     * but instead signs and encodes the properties.
     *
     * @return string
     */
    public function __toString()
    {
        return $this->getSelfSignedRequest($this->appSecret, $this->decodedPayload);
    }

    /**
     * Whether or not the request is within a sane window.
     *
     * @param integer   $signatureWindowSize
     * @param \DateTime $issuedAt
     *
     * @throws \InvalidArgumentException
     *
     * @return boolean
     */
    protected function isWithinWindow($signatureWindowSize, \DateTime $issuedAt)
    {
        if (! $issuedAt instanceof ExtendedDateTime) {
            throw new \InvalidArgumentException('Invalid timestamp supplied.');
        }

        $now = new ExtendedDateTime();

        $diff = $now->getTimestamp() - $issuedAt->getTimestamp();

        // The diff must be less than, or equal to the window size. To protect against overflow possibilities
        // we test if the differences is equal to, or greater than 0.
        if ($diff <= $signatureWindowSize && $diff >= 0) {
            return true;
        }

        return false;
    }

    /**
     * Sign the remote payload with the local (shared) secret. The result should be identical to the one we got
     * from the server.
     *
     * @param string $secret
     * @param array  $requestParameters
     *
     * @return string
     */
    protected function getSelfSignedRequest($secret, array $requestParameters)
    {
        $requestParameters['signature'] = $this->getSignatureFromParameters($secret, $requestParameters);

        return http_build_query($requestParameters);
    }

    /**
     * Generate the signature, based on the request parameters
     *
     * @param string $secret
     * @param array  $requestParameters
     *
     * @return string
     */
    protected function getSignatureFromParameters($secret, array $requestParameters)
    {
        unset($requestParameters['signature']);

        return base64_encode(
            hash_hmac(
                'sha256',
                http_build_query($requestParameters),
                $secret,
                false
            )
        );
    }

    /**
     * Validate the existence of the payload properties.
     *
     * @param array $payloadProperties
     *
     * @return bool
     */
    protected function isValidPayload(array $payloadProperties)
    {
        $defaultPayload = array(
            'appData' => null,
            'issuedAt' => null,
            'locale' => null,
            'networkEID' => null,
            'userEID' => null,
            'signature' => null
        );

        return count($defaultPayload) <= count(array_intersect_key($defaultPayload, $payloadProperties));
    }

    /**
     * Decode the payload string and convert it to an associative array.
     *
     * @param string $payload
     *
     * @return array
     */
    protected function decodePayloadToArray($payload)
    {
        parse_str($payload, $payloadProperties);

        return $payloadProperties;
    }

    /**
     * Decodes a RFC3986 encoded signed request payload
     *
     * @param string $payload
     *
     * @return string
     */
    protected function decodePayload($payload)
    {
        return urldecode($payload);
    }

    /**
     * Encode a string conform RFC3986.
     *
     * @param $payload
     *
     * @return string
     */
    protected function encodePayload($payload)
    {
        return urlencode($payload);
    }
}

File: dotnet/SpeakapAPI.NET/SignedRequest.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Cryptography;
using System.Text;

namespace SpeakapAPI
{
	/// <summary>
	/// http://developers.speakap.io/portal/tutorials/signed_request.html
	/// </summary>
	public class SignedRequest
	{
		/// <summary>
		/// The time window in seconds
		/// </summary>
		private readonly long _signatureWindowSize;

		private readonly HMACSHA256 _hmac;

		/// <summary>
		/// 
		/// </summary>
		/// <param name="appSecret"></param>
		public SignedRequest(string appSecret)
			: this(appSecret, R.DefaultSignatureWindowSize)
		{
		}

		/// <summary>
		/// The time in seconds should be added to the remote server time in order to be synchronized with the Speapap server time. Default is 0.
		/// </summary>
		/// <example>
		/// If Speapap server time is 10:00:00, and the remote server time is 10:00:20, then the value should be -20.
		/// </example>
		public double TimeDiff { get; set; }

		/// <summary>
		/// 
		/// </summary>
		/// <param name="appSecret"></param>
		/// <param name="signatureWindowSize">Time in seconds that the window is considered valid</param>
		public SignedRequest(string appSecret, long signatureWindowSize)
		{
			if (appSecret == null)
				throw new ArgumentNullException("appSecret");

			_signatureWindowSize = signatureWindowSize;
			_hmac = new HMACSHA256(Encoding.UTF8.GetBytes(appSecret));

			TimeDiff = 0;
		}

		/// <summary>
		/// request parameters in dictionary&lt;string,string&gt;
		/// </summary>
		/// <param name="parameters"></param>
		public void ValidateSignature(IDictionary<string, string> parameters, bool isWithinWindow = true)
		{
			if (parameters == null)
				throw new ArgumentNullException("parameters");

			if (!parameters.ContainsKey(R.Signature))
				throw new SpeakapSignatureValidationException("Parameters did not include a signature");

			if (!IsValidRequestParameters(parameters))
				throw new SpeakapSignatureValidationException(string.Format("Missing required parameters, got: {0}", GetSignedRequest(parameters)));

			if (parameters[R.Signature] != GetSignatureFromParameters(parameters))
				throw new SpeakapSignatureValidationException(string.Format("Invalid signature."));

			if (!isWithinWindow)
				return;

			var issuedAtUtc = DateTime.Parse(parameters[R.IssuedAt], null, System.Globalization.DateTimeStyles.RoundtripKind).ToUniversalTime();

			if (!IsWithinWindow(_signatureWindowSize, issuedAtUtc.AddSeconds(-TimeDiff)))
			{
				throw new SpeakapSignatureValidationException("Expired signature.");
			}
		}

		/// <summary>
		/// 
		/// </summary>
		/// <param name="signedRequest"></param>
		public void ValidateSignature(string signedRequest)
		{
			if(signedRequest == null)
				throw new ArgumentNullException("signedRequest");

			var parameters = SignedRequestHelper.GetParametersFromSignedRequest(signedRequest);

			ValidateSignature(parameters, false);
		}


		/// <summary>
		/// Get the encoded value. To be used in e.g. the Speakap JavaScript proxy
		/// </summary>
		/// <param name="parameters"></param>
		/// <returns></returns>
		public string GetSignedRequest(IDictionary<string, string> parameters)
		{
			if (parameters == null)
				throw new ArgumentNullException("parameters");

			if (parameters.Count == 0)
				throw new ArgumentException("parameters cannot be empty.");

			if (!parameters.ContainsKey(R.Signature))
				parameters.Add(R.Signature, string.Empty);

			parameters[R.Signature] = GetSignatureFromParameters(parameters);

			return ParametersToQueryString(parameters);
		}


		/// <summary>
		/// Get the parameters, including the signature
		/// </summary>
		/// <param name="parameters"></param>
		/// <returns></returns>
		public IDictionary<string, string> GetSignedParameters(IDictionary<string, string> parameters)
		{
			if (parameters == null)
				throw new ArgumentNullException("parameters");

			if (parameters.Count == 0)
				throw new ArgumentException("parameters cannot be empty.");

			if (!parameters.ContainsKey(R.Signature))
				parameters.Add(R.Signature, string.Empty);
			parameters[R.Signature] = GetSignatureFromParameters(parameters);

			return parameters;
		}

		/// <summary>
		/// Whether or not the request is within a sane window.
		/// </summary>
		/// <remarks>
		/// Check the absolute Time difference to overcome the local machine time is earlier than the Speakap remote server time
		/// </remarks>
		/// <param name="signatureWindowSize">signature window size in seconds</param>
		/// <param name="issuedAtUtc">UTC date time</param>
		/// <returns></returns>
		protected static bool IsWithinWindow(long signatureWindowSize, DateTime issuedAtUtc)
		{
			var diff = (DateTime.UtcNow - issuedAtUtc).TotalSeconds;
			return 0 <= diff && diff <= signatureWindowSize;
		}

		/// <summary>
		/// Sign the remote payload with the local (shared) secret. The result should be identical to the one we got from the server.
		/// </summary>
		/// <param name="parameters"></param>
		/// <returns></returns>
		protected string GetSelfSignedRequest(IDictionary<string, string> parameters)
		{
			if (parameters == null)
				throw new ArgumentNullException("parameters");

			if (!parameters.ContainsKey(R.Signature))
				parameters.Add(R.Signature, string.Empty);

			parameters[R.Signature] = GetSignatureFromParameters(parameters);

			return ParametersToQueryString(parameters);
		}

		/// <summary>
		/// Generate the signature, based on the request parameters
		/// </summary>
		/// <param name="parameters"></param>
		/// <returns></returns>
		protected string GetSignatureFromParameters(IDictionary<string, string> parameters)
		{
			if (parameters == null)
				throw new ArgumentNullException("parameters");

			if (parameters.Count == 0)
				throw new ArgumentException("parameters cannot be empty.");

			var queryString = ParametersToQueryString((from param in parameters
													   where !string.Equals(R.Signature, param.Key)
													   select param).ToDictionary(p => p.Key, p => p.Value));

			if (queryString == null)
				throw new ArgumentException("queryString");

			var inArray = _hmac.ComputeHash(Encoding.UTF8.GetBytes(queryString));

			if (inArray == null)
				throw new ArgumentException("inArray");

			return Convert.ToBase64String(inArray);
		}

		/// <summary>
		/// Validate the existence of the payload properties.
		/// </summary>
		/// <param name="parameters"></param>
		/// <returns></returns>
		protected bool IsValidRequestParameters(IDictionary<string, string> parameters)
		{
			if (parameters == null)
				throw new ArgumentNullException("parameters");

			return R.DefaultKeys.Length <= parameters.Keys.Intersect(R.DefaultKeys, StringComparer.InvariantCultureIgnoreCase).Count();
		}

		/// <summary>
		/// Convert an array to a query-string, RFC3986 encoded
		/// </summary>
		/// <param name="parameters"></param>
		/// <returns></returns>
		public static string ParametersToQueryString(IDictionary<string, string> parameters)
		{
			if (parameters == null)
				throw new ArgumentNullException("parameters");

			var queryString = string.Join(R.Separator, from param in parameters
													   where param.Key != R.Signature
													   orderby param.Key
													   select string.Format("{0}={1}", Uri.EscapeDataString(param.Key),
																					   Uri.EscapeDataString(param.Value == null ? string.Empty : param.Value)));

			if (parameters.ContainsKey(R.Signature))
			{
				var signature = string.Concat(R.Signature, "=", Uri.EscapeDataString(parameters[R.Signature]));

				queryString = string.IsNullOrEmpty(queryString) ? signature : string.Concat(queryString, R.Separator, signature);
			}

			return queryString;
		}

	}
}